torch_geometric/nn/conv/point_transformer_conv.py

```diff
@@ -0,0 +1,110 @@
+        out_channels (int) : size of each output sample features
+        add_self_loops (bool, optional): If set to :obj:`False`, will not add
+            self-loops to the input graph. (default: :obj:`True`)
+
+        pos_mlp_channels : (int, optional) size of intermediary channels in
+            points relative spatial coordinates computation.
+            (default: :obj:`64`)
+        attn_mlp_channels : (int, optional) factor size of intermediary
+            channels in transformed features mlp. (default: :obj:`2`)
+
+        **kwargs (optional): Additional arguments of
+            :class:`torch_geometric.nn.conv.MessagePassing`.
+    """
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 add_self_loops=True,
+                 pos_mlp_channels=64,
+                 attn_mlp_channels=2):
+        super(PointTransformerConv,
+              self).__init__(aggr='add')  # "Add" aggregation
+
+        self.delta = Sequential(Linear(3, pos_mlp_channels), ReLU(),
+                                Linear(pos_mlp_channels, out_channels))
+        self.gamma = Sequential(
+            Linear(out_channels, out_channels * attn_mlp_channels), ReLU(),
+            Linear(out_channels * attn_mlp_channels, out_channels))
+        self.alpha = Linear(in_channels, out_channels)
+        self.phi = Linear(in_channels, out_channels)
+        self.psi = Linear(in_channels, out_channels)
+
+        self.add_self_loops = add_self_loops
+
+    def reset_parameters(self):
+        reset(self.delta)
+        reset(self.gamma)
+        reset(self.alpha)
+        reset(self.phi)
+        reset(self.psi)
+
+    def forward(self, x: Union[OptTensor, PairOptTensor],
+                pos: Union[Tensor, PairTensor], edge_index: Adj) -> Tensor:
+        """"""
+
+        if not isinstance(x, tuple):
+            x: PairOptTensor = (x, x)
+
+        if isinstance(pos, Tensor):
+            pos: PairTensor = (pos, pos)
+
+        if self.add_self_loops:
+            if isinstance(edge_index, Tensor):
+                edge_index, _ = remove_self_loops(edge_index)
+                edge_index, _ = add_self_loops(edge_index,
+                                               num_nodes=pos[1].size(0))
+            elif isinstance(edge_index, SparseTensor):
+                edge_index = set_diag(edge_index)
+
+        return self.propagate(edge_index, x=x, pos=pos)
+
+    def message(self, x_i, x_j, pos_i, pos_j):
+
+        # query is x_i, key is x_j, value is the substraction of the two
+        out_delta = self.delta((pos_j - pos_i))  # compute position encoding
+
+        out_phi_psi = self.phi(x_i) - self.psi(
+            x_j)  # compute point wise transformation
+        out_gamma = self.gamma(out_phi_psi +
+                               out_delta)  # compute importance factor
+
+        # normalize the result with softmax and multiply
+        # by features transformed by alpha
+        msg = softmax(out_gamma, 1) * (
```

Comment:
Correct me if Iam wrong. Currently the `softmax` operation is applied over all nodes. Shouldn't be applied only over neighboring nodes?
In which case, use `softmax` from `torch_geometric.utils.softmax`, like its used in [GATConv](https://github.com/pyg-team/pytorch_geometric/blob/764ceec99698120e8a667e30625825de14ee5126/torch_geometric/nn/conv/gat_conv.py#L191).

Reply:
I think this observation is correct. The `softmax` also is currently applied in the feature dimension.

Reply:
My bad, I didn't understand correctly how x_i and x_j were related I will correct it